File: backend/otel.py
import os
from opentelemetry import trace
from opentelemetry.sdk.trace import TracerProvider
from opentelemetry.sdk.trace.export import BatchSpanProcessor
from opentelemetry.exporter.otlp.proto.grpc.trace_exporter import OTLPSpanExporter
from opentelemetry.sdk.resources import Resource

SIGNOZ_ENDPOINT = os.getenv("SIGNOZ_ENDPOINT", "https://ingest.us.signoz.cloud:443")
SIGNOZ_INGESTION_KEY = os.getenv("SIGNOZ_INGESTION_KEY", "")
SERVICE_NAME = os.getenv("SIGNOZ_SERVICE_NAME", "agent-rca-backend")

def init_otel():
    if not SIGNOZ_INGESTION_KEY:
        return None
    try:
        resource = Resource.create({"service.name": SERVICE_NAME})
        provider = TracerProvider(resource=resource)
        exporter = OTLPSpanExporter(
            endpoint=SIGNOZ_ENDPOINT,
            headers={"signoz-ingestion-key": SIGNOZ_INGESTION_KEY},
        )
        processor = BatchSpanProcessor(exporter)
        provider.add_span_processor(processor)
        trace.set_tracer_provider(provider)
        return trace.get_tracer(__name__)
    except Exception as e:
        print(f"[otel] Warning: Failed to init tracing: {e}")
        return None

File: frontend/src/otel.js
import { WebTracerProvider } from '@opentelemetry/sdk-trace-web';
import { OTLPTraceExporter } from '@opentelemetry/exporter-trace-otlp-http';
import { DocumentLoadInstrumentation } from '@opentelemetry/instrumentation-document-load';
import { FetchInstrumentation } from '@opentelemetry/instrumentation-fetch';
import { ZoneContextManager } from '@opentelemetry/context-zone';
import { Resource } from '@opentelemetry/resources';

const SIGNOZ_ENDPOINT = import.meta.env.VITE_SIGNOZ_ENDPOINT || 'https://ingest.us.signoz.cloud:443';
const SIGNOZ_INGESTION_KEY = import.meta.env.VITE_SIGNOZ_INGESTION_KEY || '';

export function initWebTracing() {
  if (!SIGNOZ_INGESTION_KEY) return;
  try {
    const exporter = new OTLPTraceExporter({
      url: `${SIGNOZ_ENDPOINT}/v1/traces`,
      headers: { 'signoz-ingestion-key': SIGNOZ_INGESTION_KEY },
    });

    const provider = new WebTracerProvider({
      resource: new Resource({ 'service.name': 'agent-rca-frontend' }),
    });

    provider.addSpanProcessor(new BatchSpanProcessor(exporter));
    provider.register({ contextManager: new ZoneContextManager() });

    DocumentLoadInstrumentation.create().setTracerProvider(provider);
    FetchInstrumentation.create({
      ignoreUrls: [/signoz\.cloud/],
      propagateTraceHeaderCorsUrls: ['http://localhost:8000/api/*'],
    }).setTracerProvider(provider);
  } catch (e) {
    console.warn('[otel] Failed to init web tracing:', e);
  }
}
